File: shared/realtime/entityId.js
// Stable guest/player id → u32 entity id (shared by client pack bridge and
// server binary encoder). Pure FNV-1a — matches Afterlight.Realtime.EntityId.

export function playerEntityId(playerId) {
  if (playerId == null || playerId === '') return 0;
  const s = String(playerId);
  let h = 0x811c9dc5;
  for (let i = 0; i < s.length; i++) {
    h ^= s.charCodeAt(i);
    h = Math.imul(h, 0x01000193);
  }
  return h >>> 0;
}

File: server_elixir/lib/afterlight/realtime/entity_id.ex
defmodule Afterlight.Realtime.EntityId do
  @moduledoc """
  Stable string player/guest id → u32 entity id for the binary data plane.
  Matches `shared/realtime/entityId.js` (FNV-1a).
  """

  import Bitwise

  @fnv_offset 0x811C9DC5
  @fnv_prime 0x01000193

  @spec hash(term) :: non_neg_integer()
  def hash(player_id) when is_binary(player_id) and player_id != "" do
    player_id
    |> :binary.bin_to_list()
    |> Enum.reduce(@fnv_offset, fn byte, acc ->
      acc = bxor(acc, byte)
      band(acc * @fnv_prime, 0xFFFFFFFF)
    end)
  end

  def hash(_), do: 0
end
